File: extensions/GPIO/clients/c/ONC-RPC/Toggle_Speed.c
// MuntsOS GPIO Thin Server Toggle Speed Test

// Copyright (C)2013-2018, Philip Munts, President, Munts AM Corp.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "gpio_server_oncrpc.h"

#define SERVERNAME	argv[1]
#define LED		atoi(argv[2])
#define ITERATIONS	atoi(argv[3])

static void ShowUsage(void)
{
  fprintf(stderr, "Usage: Toggle_Speed <servername> <pin number> <iterations>\n");
  exit(1);
}

int main(int argc, char *argv[])
{
  CLIENT *clnt;
  int result;
  struct timeval start, stop;
  int i;
  double deltat, rate, cycletime;

  puts("\nMuntsOS GPIO Thin Server Toggle Speed Test\n");

  if (argc != 4) ShowUsage();

  // Create ONC/RPC client handle

  clnt = clnt_create(SERVERNAME, GPIO_SERVER_ONCRPC, GPIO_SERVER_ONCRPC_VERS, "udp");
  if (clnt == NULL)
  {
    clnt_pcreateerror("ERROR: clnt_create() failed");
    exit(1);
  }

  // Configure GPIO pins

  result = *gpio_open_1(LED, GPIO_DIRECTION_OUTPUT, 0, clnt);
  if (result < 0)
  {
    fprintf(stderr, "ERROR: %s\n", strerror(-result));
    exit(1);
  }

  // Perform GPIO toggle speed test

  printf("Performing %d GPIO writes...\n\n", ITERATIONS);

  gettimeofday(&start, NULL);

  for (i = 0; i < ITERATIONS/2; i++)
  {
    result = *gpio_write_1(LED, 0, clnt);
    if (result < 0)
    {
      fprintf(stderr, "ERROR: %s\n", strerror(-result));
      exit(1);
    }

    result = *gpio_write_1(LED, 1, clnt);
    if (result < 0)
    {
      fprintf(stderr, "ERROR: %s\n", strerror(-result));
      exit(1);
    }
  }

  gettimeofday(&stop, NULL);

  // Display results

  deltat = (stop.tv_sec*1E6 + stop.tv_usec - start.tv_sec*1E6 - start.tv_usec)/1E6;
  rate = ITERATIONS/deltat;
  cycletime = deltat/ITERATIONS;

  printf("Performed %d GPIO writes in %1.1f seconds\n", ITERATIONS, deltat);
  printf("  %1.1f iterations per second\n", rate);
  printf("  %1.1f microseconds per iteration\n", cycletime*1E6);
  exit(0);
}

File: extensions/GPIO/clients/java/ONC-RPC/Toggle_Speed/Toggle_Speed.java
// GPIO pin toggle speed test using Remote Tea ONC/RPC

// Copyright (C)2017-2018, Philip Munts, President, Munts AM Corp.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice,
//   this list of conditions and the following disclaimer.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

import java.io.*;

import com.munts.interfaces.GPIO.Pin;
import com.munts.interfaces.GPIO.Direction;

public class Toggle_Speed
{
  public static void main(String args[])
  {
    com.munts.muntsos.examples.GPIO.ONCRPC.Server s;
    int iterations;
    int i;
    long start_time;
    long stop_time;
    double deltat;
    double rate;
    double cycletime;

    System.out.println("\nMuntsOS GPIO Thin Server Toggle Speed Test\n");

    if (args.length != 3)
    {
      System.out.println("Usage: java -jar ToggleSpeed.jar <servername> " +
        "<pin number> <iterations>\n");
      System.exit(1);
    }

    // Connect to the server

    s = new com.munts.muntsos.examples.GPIO.ONCRPC.Server(args[0]);

    // Create a GPIO pin object

    Pin p = new com.munts.muntsos.examples.GPIO.ONCRPC.Pin(s,
     Integer.parseInt(args[1]), Direction.Output);

    iterations = Integer.parseInt(args[2]);

    System.out.println("Performing " + Integer.toString(iterations) +
      " GPIO writes...\n");

    // Perform GPIO toggle speed test

    start_time = System.currentTimeMillis();

    for (i = 0; i < iterations/2; i++)
    {
      p.write(true);
      p.write(false);
    }

    stop_time = System.currentTimeMillis();

    // Display results

    deltat = (stop_time - start_time)/1000.0;
    rate = iterations/deltat;
    cycletime = deltat/iterations;

    System.out.println("Performed " + Integer.toString(iterations) +
      " in " + String.format("%1.1f", deltat) + " seconds");

    System.out.println("  " + String.format("%1.1f", rate) +
      " iterations per second");

    System.out.println("  " + String.format("%1.1f", cycletime*1E6) +
      " microseconds per iteration\n");
  }
}
